ipc/provider/src/lib.rs

```diff
@@ -263,63 +263,144 @@ impl IpcProvider {
-    pub async fn leave_subnet(&self, _subnet: SubnetID, _from: Address) -> anyhow::Result<()> {
-        todo!()
+    pub async fn leave_subnet(
+        &mut self,
+        subnet: SubnetID,
+        from: Option<Address>,
+    ) -> anyhow::Result<()> {
+        let parent = subnet.parent().ok_or_else(|| anyhow!("no parent found"))?;
+        let conn = match self.connection(&parent) {
+            None => return Err(anyhow!("target parent subnet not found")),
+            Some(conn) => conn,
+        };
+
+        let subnet_config = conn.subnet();
+        self.check_subnet(subnet_config)?;
+        let sender = self.check_sender(subnet_config, from)?;
+
+        conn.manager().leave_subnet(subnet, sender).await
     }
 
-    /// Sends a signal to kill a subnet
-    pub async fn kill_subnet(&self, _subnet: SubnetID, _from: Address) -> anyhow::Result<()> {
-        todo!()
+    pub async fn kill_subnet(
+        &mut self,
+        subnet: SubnetID,
+        from: Option<Address>,
+    ) -> anyhow::Result<()> {
+        let parent = subnet.parent().ok_or_else(|| anyhow!("no parent found"))?;
+        let conn = match self.connection(&parent) {
+            None => return Err(anyhow!("target parent subnet not found")),
+            Some(conn) => conn,
+        };
+
+        let subnet_config = conn.subnet();
+        self.check_subnet(subnet_config)?;
+        let sender = self.check_sender(subnet_config, from)?;
+
+        conn.manager().kill_subnet(subnet, sender).await
     }
 
-    /// Lists all the registered children in a gateway.
     pub async fn list_child_subnets(
         &self,
-        _gateway_addr: Address,
+        gateway_addr: Option<Address>,
+        subnet: &SubnetID,
     ) -> anyhow::Result<HashMap<SubnetID, SubnetInfo>> {
-        todo!()
+        let parent = subnet.parent().ok_or_else(|| anyhow!("no parent found"))?;
+        let conn = match self.connection(&parent) {
+            None => return Err(anyhow!("target parent subnet not found")),
+            Some(conn) => conn,
+        };
+
+        let subnet_config = conn.subnet();
+        self.check_subnet(subnet_config)?;
+
+        let gateway_addr = match gateway_addr {
+            None => subnet_config.gateway_addr(),
+            Some(addr) => addr,
+        };
+
+        conn.manager().list_child_subnets(gateway_addr).await
     }
 
-    /// Fund injects new funds from an account of the parent chain to a subnet.
-    /// Returns the epoch that the fund is executed in the parent.
     pub async fn fund(
-        &self,
-        _subnet: SubnetID,
-        _gateway_addr: Address,
-        _from: Address,
-        _to: Address,
-        _amount: TokenAmount,
+        &mut self,
+        subnet: SubnetID,
+        gateway_addr: Option<Address>,
+        from: Option<Address>,
+        to: Address,
+        amount: TokenAmount,
     ) -> anyhow::Result<ChainEpoch> {
-        todo!()
+        let parent = subnet.parent().ok_or_else(|| anyhow!("no parent found"))?;
```

Comment:
maybe we should remove this as well, so that we can fund address in `root` subnets

Reply:
I don't think this is the case, right? When funding you need to connect to the parent and send a `fund` to the gateway. So there shouldn't be a `fund` from the root as an argument.